contrib/codegen/src/database.rs

```diff
@@ -118,8 +118,8 @@ pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStrea
             /// Retrieves a connection of type `Self` from the `rocket`
             /// instance. Returns `Some` as long as `Self::fairing()` has been
             /// attached and there is at least one connection in the pool.
-            pub fn get_one(rocket: &::rocket::Rocket) -> Option<Self> {
-                rocket.state::<#pool_type>()
+            pub fn get_one(manifest: &::rocket::Manifest) -> Option<Self> {
+                manifest.state::<#pool_type>()
```

Comment:
Actually, taking an `&mut Rocket` should be (mostly) fine:

```rust
let mut rocket = make_rocket();
let conn = rocket.get_one().await;
let config = rocket.config().await; // <-- The code I was afraid would not work, but is actually fine.
```

Since `get_one()` returns owned data, there is no conflict between calling `get_one` and other methods.

However, I see at least one place we would "lose" something: `on_launch` will no longer be able to call `get_one`. And `on_launch` shouldn't change to take `&mut Rocket`, because by that point it is too late to make many of the modifications that `&mut Rocket` allows.

Reply:
That's a good point.

I suppose one idea, perhaps a bad one, is to use some form of interior mutability - say a reentrant lock - to make `inspect()` take an `&Rocket` instead of an `&mut`. I believe if we did that, however, we wouldn't need to expose `Manifest` at all, which is kind of interesting. In some ways, this feels right as `.inspect()` is logically read-only, and only takes an `&mut` due to its implementation.

What do you think about this? Are we gaining anything by separating `Rocket` and `Manifest`?

Reply:
> say a reentrant lock

I think I can see it in principle, but it has some *really* awkward consequences:

```rust
let conf = rocket.config().await;                          // so far so good
let rocket = rocket.attach(AdHoc::on_attach(|rocket|
    Ok(rocket::ignite().manage(X))
));                                                        // this attach fairing is "queued"
let state = rocket.state::<X>().await;                     // this must run the attach fairing

conf.get_something();                                      // oops
```

Effectively, `inspect()` or functions that call it would now have to panic if outstanding borrows exist - this also includes calls made from `on_launch()`.